store/store.go

```diff
@@ -0,0 +1,322 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"gopkg.in/macaroon.v1"
+
+	"github.com/CanonicalLtd/serial-vault/datastore"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/overlord/auth"
+)
+
+const (
+	ssoBaseURL   = "https://login.ubuntu.com/api/v2/"
+	storeBaseURL = "https://dashboard.snapcraft.io/dev/api/"
+)
+
+// Permissions is the SSO authorization for the store
+type Permissions struct {
+	Permissions []string `json:"permissions"`
+}
+
+// ACL is the SSO authorization for the store
+type ACL struct {
+	Macaroon string `json:"macaroon"`
+}
+
+// Discharge is the SSO authorization for the store
+type Discharge struct {
+	Macaroon string `json:"discharge_macaroon"`
+}
+
+// Auth is the SSO authorization for the store
+type Auth struct {
+	Email    string `json:"email"`
+	Password string `json:"password"`
+	OTP      string `json:"otp"`
+}
+
+// KeyRegister is the request to submit a signing-key to the store
+type KeyRegister struct {
+	Auth
+	AuthorityID string `json:"authority-id"`
+	KeyName     string `json:"key-name"`
+}
+
+// RegisterKey registers an account-key assertion with the store
+func RegisterKey(keyAuth KeyRegister, keypair datastore.Keypair) error {
+
+	// Login to the store as the user
+	// Permissions are documented at https://dashboard.snapcraft.io/docs/api/macaroon.html#request-a-macaroon
+	permissions := []string{"edit_account", "modify_account_key"}
+	m, discharge, err := LoginUser(keyAuth.Email, keyAuth.Password, keyAuth.OTP, permissions)
+	if err != nil {
+		log.Println("Error logging in to store", err)
+		return err
+	}
+
+	// Generate the authorization header from the macaroons
+	authHeader, err := AuthorizationHeader(m, discharge)
+	if err != nil {
+		return err
+	}
+
+	// Generate the account-key request assertion
+	assert, err := generateAccountKeyRequest(keyAuth, keypair)
+	if err != nil {
+		return err
+	}
+
+	// The assertion needs to be sent as JSON (undocumented)
+	data := map[string]string{
+		"account_key_request": string(assert),
+	}
+	d, err := json.Marshal(data)
+	if err != nil {
+		log.Printf("Error marshalling account-key assertion: %v", err)
+		return err
+	}
+
+	// Submit the account-key assertion
+	r, _ := http.NewRequest("POST", storeBaseURL+"account/account-key", bytes.NewReader(d))
+	r.Header.Set("Authorization", authHeader)
+	r.Header.Set("Content-Type", "application/json")
+	r.Header.Set("Accept", "application/json")
+	client := http.Client{}
+	_, err = client.Do(r)
+	if err != nil {
+		log.Printf("Error submitting the account-key assertion: %v", err)
+		return err
+	}
+
+	// defer resp.Body.Close()
+
+	// b, err := ioutil.ReadAll(resp.Body)
+	// log.Println(string(b))
```

Comment:
Why this commented out code?

Reply:
That's leftover debugging - it was a long painful journey to get to this solution!